TDD for UaaTokenStore: count DB usage in performExpirationClean

strehle opened the pull request:
Add test to check the problem which is that performExpirationClean is called to often

strehle committed:
test only: count performExpirationClean

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into test/performExpirationCleanCounter

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -332,6 +337,32 @@ void beOAuth2StandardCompliant() {
+    @Test
+    void testCountingTheExecutedSqlDeleteStatements() throws SQLException {
+        // Given, mocked data source to count how often it is used, call performExpirationClean 10 times
+        DataSource mockedDataSource = mock(DataSource.class);
+        Connection mockedConnection = mock(Connection.class);
+        doReturn(mockedConnection).when(mockedDataSource).getConnection();
+        store = new UaaTokenStore(mockedDataSource);
```
I spent 15 minutes staring at this code to understand what it does, but I just can't. I don't understand why you are mocking the `getConnection()` method. I can't even see it used in the `UaaTokenStore` code. 

It looks to me like you are doing some quite low level white box testing. I would much rather try to write high level tests.

strehle replied:
see now, just close this comment as it is obsolete

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -332,6 +337,32 @@ void beOAuth2StandardCompliant() {
+            }
+        }
+        // Then
+        verify(mockedConnection, atMost(1)).prepareStatement("delete from oauth_code where expiresat > 0 AND expiresat < ?");
```
One idea: it seems to me like we could for example change the signature of the `performExpirationClean` method, to return a boolean. It would be `true` if the cleanup happened, and `false` otherwise. 
That way, the assertions would be much easier to write, wouldn't they?

strehle replied:
see now, just close this comment as it is obsolete

bruce-ricard reviewed (changes requested):
I'm not comfortable accepting this, as I don't understand the code. But thank you very much for writing a test for this code.

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into test/performExpirationCleanCounter

strehle committed:
simplify test and verification

run 10 times, without extra call
SQL itself not needed to be verified, but check on usage of getConnection

strehle committed:
Merge branch 'develop' of github.com:cloudfoundry/uaa into test/performExpirationCleanCounter
